Introduce validator index syncer

dknopik opened the pull request:
- Switch index to Option
- Clean up event syncer Config
- Remove obsolete testing binary

dknopik committed:
Introduce validator index syncer

- Switch index to Option
- Clean up syncer event Config
- Remove obsolete testing binary

dknopik committed:
Merge branch 'unstable' into index-in-database

dknopik committed:
remove accidental dep

dknopik committed:
do not crash syncer on API error and use POST instead of GET

dknopik committed:
fix SQL

dknopik committed:
do a sweep instead of a shuffle and only get unliquidated clusters

dknopik committed:
fix code

dknopik committed:
cargo fmt

Zacholme7 commented on anchor/client/src/cli.rs:
```diff
@@ -111,6 +111,15 @@ pub struct Node {
+                Default is http://localhost:8546.",
+        display_order = 0
+    )]
+    pub execution_nodes_websocket: Option<Vec<String>>,
```
nits on these. 
1) help is calling execution endpoints beacon endpoints
2) Maybe simplify these to execution_http & execution_ws?

dknopik replied:
Decided on `execution_rpc` instead of `execution_http` to emphasise the difference, is that fine?

Zacholme7 replied:
yep, thanks

Zacholme7 commented on anchor/database/src/validator_operations.rs:
```diff
@@ -61,4 +63,38 @@ impl NetworkDatabase {
+                        .multi_state
+                        .validator_metadata
+                        .update(&public_key, validator);
+                }
```
Maybe warn/error log missing validators? Should the entries always exist?

dknopik replied:
There is the edge case of a validator being removed while the request is in flight. That is not really an error though - so I decided against logging. But given the rarity of that edge case a warn log might be worth it, yeah

Zacholme7 commented on anchor/eth/src/index_sync.rs:
```diff
@@ -0,0 +1,144 @@
+const INDEX_SYNCER_NAME: &str = "validator_index_syncer";
+
+const MAX_BATCH_SIZE: usize = 512;
+const BATCHING_DELAY: Duration = Duration::from_secs(1);
```
Think 1second is large enough? sig verify is the bottle neck in processing and this coincides with validators added. On a heavy block with lots of batched validator added events it might be useful to have more of a buffer on this. Fetching from the bn is the quick part.

dknopik replied:
The delay resets if a validator is fetched from the queue - so it's second per validator, which should be fine. my tests on interop showed that we regularly hit the max batch size during sync from the queue alone

I also do not want to increase the delay too much as we want to get the index ASAP during live operation.

Zacholme7 replied:
Ah okay, nice.

Zacholme7 reviewed (commented):
This is slick, nice job. Left tiny nits

Zacholme7 commented on anchor/eth/src/index_sync.rs:
```diff
@@ -0,0 +1,144 @@
+        // next, fill up the rest of the batch with older validators that are unknown from the
+        // database
+        let space = MAX_BATCH_SIZE - batch.len();
+        if space > 0 {
+            let state = db.state();
+            let clusters = state.clusters();
+            let mut from_database = state
+                .metadata()
+                .values()
+                .filter_map(|v| {
+                    (v.index.is_none()
+                        && !batch.contains(&ValidatorId::PublicKey(v.public_key))
+                        && clusters
+                            .get_by(&v.cluster_id)
+                            .is_some_and(|c| !c.liquidated))
+                    .then_some(v.public_key)
+                })
+                .collect::<Vec<_>>();
+            drop(state);
+            let count = from_database.len();
+            debug!(len = count, db_sweep, "Found unset index validators");
+
+            // sort and skip to current position
+            from_database.sort_unstable_by_key(|x| x.serialize());
+            batch.extend(
+                from_database
+                    .into_iter()
+                    .skip(db_sweep)
+                    .take(space)
+                    .map(ValidatorId::PublicKey),
+            );
```
I think this could all be turned into a quick db query. Something like 
`
SELECT public_key 
FROM validators 
WHERE index IS NULL;
`
not sure if this would result in a speedup or not though

dknopik replied:
pushed a test, lets try it out :)

Will report back

dknopik replied:
It's slower by a factor of ~50 (which surprised me), therefore I propose leaving it like this

Zacholme7 replied:
Oh interesting, I suppose its still fetching from db at the end of the day. thanks for giving it a shot

dknopik committed:
properly handle index conflict

dknopik committed:
better names for CLI options

dknopik committed:
warn on unknown validator index update

dknopik committed:
test out getting candidates from db instead

dknopik committed:
Merge branch 'unstable' into index-in-database

dknopik committed:
Revert "test out getting candidates from db instead"

This reverts commit a92aa262a785b5244b013cf19c29b089adc0b0af.

diegomrsantos commented on anchor/client/src/config.rs:
```diff
@@ -36,6 +36,8 @@ pub struct Config {
     pub proposer_nodes: Vec<SensitiveUrl>,
     /// The http endpoints of the execution node APIs.
     pub execution_nodes: Vec<SensitiveUrl>,
+    /// The websocket endpoints of the execution node APIs.
+    pub execution_nodes_websocket: Vec<SensitiveUrl>,
```
nit: maybe group them in a struct?

dknopik replied:
Hmm, I think personally I prefer it flat here. Do you have specific advantages in mind that I might not see right now?

diegomrsantos replied:
something like that seems nicer and easier to pass around:

```rust
pub struct ExecutionNodes {
    pub http: Vec<SensitiveUrl>,
    pub websocket: Vec<SensitiveUrl>,
}

pub struct NodesConfig {
    pub beacon_nodes: Vec<SensitiveUrl>,
    pub proposer_nodes: Vec<SensitiveUrl>,
    pub execution_nodes: ExecutionNodes,
}
```

diegomrsantos replied:
But maybe a matter of taste

Zacholme7 replied:
I also prefer it flat

diegomrsantos commented on anchor/client/src/config.rs:
```diff
@@ -132,14 +133,22 @@ pub fn from_cli(cli_args: &Node) -> Result<Config, String> {
             .map_err(|e| format!("Unable to parse execution node URL: {:?}", e))?;
     }
 
+    if let Some(execution_nodes_websocket) = &cli_args.execution_ws {
+        config.execution_nodes_websocket = execution_nodes_websocket
+            .iter()
+            .map(|s| SensitiveUrl::parse(s))
+            .collect::<Result<_, _>>()
+            .map_err(|e| format!("Unable to parse execution node URL: {:?}", e))?;
+    }
+
```
It'd be nice to remove the duplication with the code above

diegomrsantos commented on anchor/common/ssv_types/src/cluster.rs:
```diff
@@ -77,7 +77,7 @@ pub struct ValidatorMetadata {
     pub cluster_id: ClusterId,
     /// Index of the validator
-    pub index: ValidatorIndex,
+    pub index: Option<ValidatorIndex>,
```
Could you elaborate on the motivation behind this?

dknopik replied:
We do not always have a validator index, as it is not supplied via SSV contract event. There, the validator public key is supplied instead.

The validator only gets a validator index once it receives a beacon chain deposit. This means someone might enable a validator in SSV, but the validator is not active in the beacon chain. So we have no validator index for that validator (and might never have, actually!).

Before the change, we represented this scenario by setting a validator index of 0. This is not idiomatic vs just representing the absence of a value via an `Option`, which is why I changed it.

diegomrsantos replied:
> This means someone might enable a validator in SSV, but the validator is not active in the beacon chain

Is there a way to allow validator registration only if it's active in the beacon chain?

dknopik replied:
I think it might be possible if you pass a giant merkle proof over the beacon state to the SSV contract :D It's definitely not worth it.

There is also the use-case of registering a validator at the same time as depositing it - the validator is not immediately active then, but will become active soon.

dknopik committed:
remove duplication

dknopik committed:
Merge branch 'unstable' into index-in-database

diegomrsantos commented on anchor/client/src/config.rs:
```diff
@@ -236,6 +237,22 @@ pub fn from_cli(cli_args: &Node) -> Result<Config, String> {
 }
 
+/// Read SensitiveUrls from given CLI Strings
+fn read_urls(
```
maybe `sanitize_urls`?

diegomrsantos commented on anchor/client/src/config.rs:
```diff
@@ -127,21 +128,21 @@ pub fn from_cli(cli_args: &Node) -> Result<Config, String> {
-            .collect::<Result<_, _>>()
-            .map_err(|e| format!("Unable to parse execution node URL: {:?}", e))?;
-    }
+    read_urls(
+        &mut config.beacon_nodes,
+        &cli_args.beacon_nodes,
+        "beacon node",
+    )?;
```
```suggestion
if let Some(ref beacon_nodes) = cli_args.beacon_nodes {
    read_urls(&mut config.beacon_nodes, beacon_nodes, "beacon node")?;
}
```

dknopik replied:
I wanted to minimise repetition, which was the purpose of this change

diegomrsantos replied:
Imo readability of the code is also important. It's more intuitive to not call the function if the input is none.

diegomrsantos replied:
Another option is grouping the configs in something like this:

```rust
pub enum NodeType {
    Beacon,
    Proposer,
    Execution(ExecutionType),
}

pub enum ExecutionType {
    Http,
    WebSocket,
}

pub struct NodeEndpoint {
    pub node_type: NodeType,
    pub url: SensitiveUrl,
}

pub struct NodesConfig {
    pub endpoints: Vec<NodeEndpoint>,
}
```

But I don't know if it'd make other things more difficult

diegomrsantos replied:
Or even simpler

```rust
pub enum Node {
    Beacon(SensitiveUrl),
    Proposer(SensitiveUrl),
    ExecutionHttp(SensitiveUrl),
    ExecutionWebSocket(SensitiveUrl),
}

pub struct NodesConfig {
    pub nodes: Vec<Node>,
}
```

dknopik replied:
Then we have to filter out the appropriate URLs later on, again creating Vecs of the filtered values :sweat_smile: really not a fan of this

diegomrsantos replied:
It's not that bad to add a `get_beacon_nodes` function to NodesConfig, this `vec` is pretty small. But I see you don't like it, and I don't have a strong opinion at all.

diegomrsantos commented on anchor/client/src/config.rs:
```diff
@@ -236,6 +237,22 @@ pub fn from_cli(cli_args: &Node) -> Result<Config, String> {
+    }
+    Ok(())
+}
+
```
```suggestion
fn read_urls(
    dest: &mut Vec<SensitiveUrl>,
    src: &[String],
    kind: &str,
) -> Result<(), String> {
    *dest = src
        .iter()
        .map(|s| SensitiveUrl::parse(s))
        .collect::<Result<_, _>>()
        .map_err(|e| format!("Unable to parse {kind} URL: {:?}", e))?;
    Ok(())
}

```

diegomrsantos commented on anchor/client/src/lib.rs:
```diff
@@ -322,25 +324,26 @@ impl Client {
         wait_for_genesis(&beacon_nodes, genesis_time).await?;
 
+        // Start validator index syncer
+        let index_sync_tx =
```
It's unrelated to this PR, but could we split the `run` function into smaller ones? It's almost 500 lines long, way too big, IMO.

dknopik committed:
further improve `from_cli`

Co-authored-by: diegomrsantos <diegomrsantos@gmail.com>

diegomrsantos commented on anchor/database/src/validator_operations.rs:
```diff
@@ -62,4 +66,40 @@ impl NetworkDatabase {
+        for (public_key, index) in map.iter() {
+            transaction
+                .prepare_cached(SQL[&SqlStatement::SetIndex])?
+                .execute(params![
```
how big can the map get?

dknopik replied:
Up to `MAX_BATCH_SIZE`, which is 512.

diegomrsantos commented on anchor/eth/src/event_processor.rs:
```diff
@@ -8,13 +8,21 @@ use tracing::{debug, error, info, instrument, trace, warn};
 type EventHandler = fn(&EventProcessor, &Log) -> Result<(), ExecutionError>;
 
+pub enum Mode {
+    Node {
```
Could we add a comment describing what this mode means?

diegomrsantos commented on anchor/eth/src/event_processor.rs:
```diff
@@ -23,12 +31,12 @@ pub struct EventProcessor {
     pub db: Arc<NetworkDatabase>,
     /// Signal if we should only do relevant keysplitting processing
-    keysplit: bool,
+    mode: Mode,
```
Should we update the comment?

dknopik replied:
No, the purpose is still the same

diegomrsantos commented on anchor/eth/src/event_processor.rs:
```diff
@@ -235,9 +239,12 @@ impl EventProcessor {
 
         // During keysplitting, we only care about the nonce
-        if self.keysplit {
+        let Mode::Node {
```
Should we update the comment?

dknopik replied:
No, the purpose is still the same

diegomrsantos commented on anchor/eth/src/index_sync.rs:
```diff
@@ -0,0 +1,145 @@
+async fn validator_index_syncer(
+    nodes: Arc<BeaconNodeFallback<impl SlotClock>>,
+    db: Arc<NetworkDatabase>,
+    mut validator_queue_rx: UnboundedReceiver<PublicKeyBytes>,
```
Should we make it bounded?

diegomrsantos commented on anchor/eth/src/index_sync.rs:
```diff
@@ -0,0 +1,145 @@
+        if space > 0 {
+            let state = db.state();
+            let clusters = state.clusters();
+            let mut from_database = state
+                .metadata()
+                .values()
+                .filter_map(|v| {
+                    (v.index.is_none()
+                        && !batch.contains(&ValidatorId::PublicKey(v.public_key))
+                        && clusters
+                            .get_by(&v.cluster_id)
+                            .is_some_and(|c| !c.liquidated))
+                    .then_some(v.public_key)
+                })
+                .collect::<Vec<_>>();
```
This is a bit convoluted for a declarative approach imo

diegomrsantos replied:
But if you'd like to keep it, I'd extract the logic inside the filter to a function

diegomrsantos commented on anchor/eth/src/index_sync.rs:
```diff
@@ -0,0 +1,145 @@
+            debug!(len = batch.len(), "Sending request");
+            let validators = nodes
+                .first_success(move |client| {
+                    let batch = batch.clone();
```
don't know how often this will be called, but can this clone get expensive?

dknopik replied:
Unfortunately, I do not see a way to avoid a clone, as `post_beacon_states_validators` wants a `Vec`. This is called once per connected beacon node.

dknopik replied:
I'll shuffle the code around a bit to also handle mapping from PublicKeyBytes to ValidatorId here, this way it looks a bit more useful ;D

dknopik committed:
Add clarifying comments

dknopik committed:
Improve flow a bit

- Use `recv_many`, which should be more performant
- Extract some logic into separate function
- Handle mapping in `first_success` closure

dknopik committed:
Switch to bounded queue

diegomrsantos reviewed (approved):
Thank you for this PR and for addressing the comments. Sorry they were too many ;)

dknopik commented:
Thank you for having a good look at my code!
